File: sdks/typescript/src/identifier.ts
import { ConfigIdentifier, Scope } from './types';

/**
 Fluent builder for ConfigIdentifier
 
  @example
  ```typescript
  const identifier = new IdentifierBuilder('my-service')
    .withScope(Scope.SYSTEM)
    .withGroupId('database')
    .build();
  ```
 */
export class IdentifierBuilder {
  private identifier: ConfigIdentifier;

  constructor(serviceName: string) {
    this.identifier = {
      serviceName,
      scope: Scope.SCOPE_UNSPECIFIED,
      groupId: '',
    };
  }

  withScope(scope: Scope): this {
    this.identifier.scope = scope;
    return this;
  }

  withGroupId(groupId: string): this {
    this.identifier.groupId = groupId;
    return this;
  }

  withProjectId(projectId: string): this {
    this.identifier.projectId = projectId;
    return this;
  }

  withStoreId(storeId: string): this {
    this.identifier.storeId = storeId;
    return this;
  }

  withUserId(userId: string): this {
    this.identifier.userId = userId;
    return this;
  }

  build(): ConfigIdentifier {
    return { ...this.identifier };
  }
}

export function createIdentifier(serviceName: string): IdentifierBuilder {
  return new IdentifierBuilder(serviceName);
}

File: sdks/go/identifier.go
package scopeconfig

import (
	configv1 "github.com/digitalsolutionsai/scope-config-service/sdks/go/gen/config/v1"
)

// IdentifierBuilder provides a fluent API for building ConfigIdentifier objects.
type IdentifierBuilder struct {
	identifier *configv1.ConfigIdentifier
}

// NewIdentifier creates a new IdentifierBuilder with the required service name.
func NewIdentifier(serviceName string) *IdentifierBuilder {
	return &IdentifierBuilder{
		identifier: &configv1.ConfigIdentifier{
			ServiceName: serviceName,
			Scope:       configv1.Scope_SCOPE_UNSPECIFIED,
		},
	}
}

// WithScope sets the scope for the configuration.
func (b *IdentifierBuilder) WithScope(scope configv1.Scope) *IdentifierBuilder {
	b.identifier.Scope = scope
	return b
}

// WithGroupID sets the group ID.
func (b *IdentifierBuilder) WithGroupID(groupID string) *IdentifierBuilder {
	b.identifier.GroupId = groupID
	return b
}

// WithProjectID sets the project ID.
func (b *IdentifierBuilder) WithProjectID(projectID string) *IdentifierBuilder {
	b.identifier.ProjectId = projectID
	return b
}

// WithStoreID sets the store ID.
func (b *IdentifierBuilder) WithStoreID(storeID string) *IdentifierBuilder {
	b.identifier.StoreId = storeID
	return b
}

// WithUserID sets the user ID.
func (b *IdentifierBuilder) WithUserID(userID string) *IdentifierBuilder {
	b.identifier.UserId = userID
	return b
}

// Build returns the constructed ConfigIdentifier.
func (b *IdentifierBuilder) Build() *configv1.ConfigIdentifier {
	return b.identifier
}
